caproto/_log.py

```diff
@@ -143,9 +154,102 @@ def color_logs(color):
 
 
 logger = logging.getLogger('caproto')
+ch_logger = logging.getLogger('caproto.ch')
+search_logger = logging.getLogger('caproto.bcast.search')
 current_handler = None  # overwritten below
 
 
+class logger_nameFilter(logging.Filter):
+    def __init__(self, logger_names):
+        self.logger_names = logger_names
+    def filter(self, record):
+        for i in self.logger_names:
+            if fnmatch.fnmatch(record.name, i):
+                return True
+        return False
+
+
+class PVFilter(logging.Filter):
+    '''
+    Pass through only messages relevant to one or more PV names and
+    pv-unrelated messages.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        True if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', exclusive=False):
+        self.names = names
+        if isinstance(level, int):
+            self.levelno = level
+        elif isinstance(level, str):
+            self.levelno = logging.getLevelName(level)
+        else:
+            raise ValueError('level should be either string or int')
+        self.exclusive = exclusive
+
+    def filter(self, record):
+        if hasattr(record, 'pv'):
+            for i in self.names:
+                if fnmatch.fnmatch(record.pv, i) and record.levelno >= self.levelno:
+                    return True
+            return False
+        else:
+            return not self.exclusive
+
+
+class AddressFilter(logging.Filter):
+    def __init__(self, address_list, level='NOTSET', exclusive=False):
+        self.address_list = address_list
+        if isinstance(level, int):
+            self.levelno = level
+        elif isinstance(level, str):
+            self.levelno = logging.getLevelName(level)
```

Comment:
Since `getLevelName` is not type-stable, we need to check that this returns an `int`. That is, we need to differentiate between these two cases:

```py
In [4]: logging.getLevelName('DEBUG')
Out[4]: 10

In [5]: logging.getLevelName('SOMETHING ELSE')
Out[5]: 'Level SOMETHING ELSE'
```

If it does not return an int, we should raise a `ValueError`. Since this pattern is used in every filter class you define, it should probably be pulled out into a separate function like:

```py
def validate_level(level):
    "Accept int or string level. Return int or raise if conversion is not possible."
    ...
    return levelno
```